lightning/src/chain/chainmonitor.rs

```diff
@@ -0,0 +1,226 @@
+			let mut monitors = self.monitors.lock().unwrap();
+			for monitor in monitors.values_mut() {
+				let mut txn_outputs = monitor.block_connected(header, txdata, height, &*self.broadcaster, &*self.fee_estimator, &*self.logger);
+				has_new_outputs_to_watch |= !txn_outputs.is_empty();
+
+				if let Some(ref chain_source) = self.chain_source {
+					for (txid, outputs) in txn_outputs.drain(..) {
+						for (idx, output) in outputs.iter().enumerate() {
+							chain_source.register_output(&OutPoint { txid, index: idx as u16 }, &output.script_pubkey);
+						}
+					}
+				}
+			}
+		}
+		has_new_outputs_to_watch
+	}
+
+	/// Dispatches to per-channel monitors, which are responsible for updating their on-chain view
+	/// of a channel based on the disconnected block. See [`ChannelMonitor::block_disconnected`] for
+	/// details.
+	///
+	/// [`ChannelMonitor::block_disconnected`]: ../channelmonitor/struct.ChannelMonitor.html#method.block_disconnected
+	pub fn block_disconnected(&self, header: &BlockHeader, disconnected_height: u32) {
+		let mut monitors = self.monitors.lock().unwrap();
+		for monitor in monitors.values_mut() {
+			monitor.block_disconnected(header, disconnected_height, &*self.broadcaster, &*self.fee_estimator, &*self.logger);
+		}
+	}
+
+	/// Creates a new `ChainMonitor` used to watch on-chain activity pertaining to channels.
+	///
+	/// When an optional chain source implementing [`chain::Filter`] is provided, the chain monitor
+	/// will call back to it indicating transactions and outputs of interest. This allows clients to
+	/// pre-filter blocks or only fetch blocks matching a compact filter. Otherwise, clients may
+	/// always need to fetch full blocks absent another means for determining which blocks contain
+	/// transactions relevant to the watched channels.
+	///
+	/// [`chain::Filter`]: ../trait.Filter.html
+	pub fn new(chain_source: Option<C>, broadcaster: T, logger: L, feeest: F) -> Self {
+		Self {
+			monitors: Mutex::new(HashMap::new()),
+			chain_source,
+			broadcaster,
+			logger,
+			fee_estimator: feeest,
+		}
+	}
+
+	/// Adds the monitor that watches the channel referred to by the given outpoint.
+	///
+	/// Calls back to [`chain::Filter`] with the funding transaction and outputs to watch.
+	///
+	/// [`chain::Filter`]: ../trait.Filter.html
+	fn add_monitor(&self, outpoint: OutPoint, monitor: ChannelMonitor<ChanSigner>) -> Result<(), MonitorUpdateError> {
+		let mut monitors = self.monitors.lock().unwrap();
+		let entry = match monitors.entry(outpoint) {
+			hash_map::Entry::Occupied(_) => return Err(MonitorUpdateError("Channel monitor for given outpoint is already present")),
+			hash_map::Entry::Vacant(e) => e,
+		};
+		{
+			let funding_txo = monitor.get_funding_txo();
+			log_trace!(self.logger, "Got new Channel Monitor for channel {}", log_bytes!(funding_txo.0.to_channel_id()[..]));
+
+			if let Some(ref chain_source) = self.chain_source {
+				chain_source.register_tx(&funding_txo.0.txid, &funding_txo.1);
```

Comment:
We can drop this now that we always put the funding txo in outputs_to_watch, correct?

Reply:
Yep, done!

Reply:
Oops! I was wrong lol. Thanks for following the full changeset through to notice that `register_tx` was now unused, which is wrong :). So the issue here is that `register_output` only implies "spends of the given output", whereas `register_tx` implies "I care about the tx I'm telling you about" - for the funding tx, indeed, we care about spends, but *also* we care about the tx. The tests (obviously, ugh) don't have coverage here, presumably because now we're relying on the in-`ChannelMonitor` filtering to catch bugs (which is almost entirely good news, fewer cases to test), and it does everything except for this one thing, where we care about the ChannelManager getting notified, not the monitor!

Reply:
Removed the fix-up commit that removed `register_tx`.

Reply:
This is still confusing, we have ChainMonitor registering the funding_tx for the account of ChannelManager. A ChannelMonitor is never interested by confirmation of the funding transaction itself, only by its _spends_.

We should either:
* pass a chain::Filter to ChannelManager which would call itself register_tx and ChainMonitor restrains to call register_output on funding
* add a new MonitorEvent::FundingConfirmed returns through `release_pending_monitor_events` and move funding confirmation tracking directly into ChannelMonitor

I think 2nd option is better as you increase fault-tolerance in a distributed deployment.

Reply:
I dont know if it changes the logic much, but the second options seems like a really great idea if only because we could potentially get rid of block_connected in ChannelManager entirely.